server/src/main/java/io/crate/types/TypeCompatibility.java

```diff
@@ -21,117 +21,53 @@
-    }
-
-    @Nullable
-    private static DataType<?> compatibility(DataType<?> fromType, DataType<?> toType) {
-        if (fromType.equals(toType)) {
-            return toType;
+    public static DataType<?> getCommonType(DataType<?> type1, DataType<?> type2) {
+        if (type1.equals(type2)) {
+            return type2;
         }
-
-        if (fromType.equals(UndefinedType.INSTANCE)) {
-            return toType;
+        if (type1.equals(UndefinedType.INSTANCE)) {
+            return type2;
         }
-
-        if (toType.equals(UndefinedType.INSTANCE)) {
-            return fromType;
+        if ((type2).equals(UndefinedType.INSTANCE)) {
+            return type1;
         }
-
-        String fromTypeBaseName = fromType.getTypeSignature().getBaseTypeName();
-        String toTypeBaseName = toType.getTypeSignature().getBaseTypeName();
-        if (fromTypeBaseName.equals(toTypeBaseName)) {
+        String type1Base = type1.getTypeSignature().getBaseTypeName();
+        String type2Base = type2.getTypeSignature().getBaseTypeName();
+        if (type1Base.equals(type2Base)) {
             // If given types share the same base, e.g. arrays, parameter types must be compatible.
-            if (fromType.getTypeParameters().isEmpty() == false || toType.getTypeParameters().isEmpty() == false) {
-                return typeCompatibilityForParametrizedType(fromType, toType);
+            if (type1.getTypeParameters().isEmpty() == false || (type2).getTypeParameters().isEmpty() == false) {
+                try {
+                    return DataTypes.merge(type1, type2);
+                } catch (IllegalArgumentException ex) {
+                    return null;
+                }
             }
-            return fromType;
+            return type1;
         }
-
-        return convertTypeByPrecedence(fromType, toType);
+        return convertTypeByPrecedence(type1, type2);
     }
 
     @Nullable
-    private static DataType<?> convertTypeByPrecedence(DataType<?> arg1, DataType<?> arg2) {
+    private static DataType<?> convertTypeByPrecedence(DataType<?> type1, DataType<?> type2) {
         final DataType<?> higherPrecedenceArg;
         final DataType<?> lowerPrecedenceArg;
-        if (arg1.precedes(arg2)) {
-            higherPrecedenceArg = arg1;
-            lowerPrecedenceArg = arg2;
+        if (type1.precedes(type2)) {
+            higherPrecedenceArg = type1;
+            lowerPrecedenceArg = type2;
         } else {
-            higherPrecedenceArg = arg2;
-            lowerPrecedenceArg = arg1;
+            higherPrecedenceArg = type2;
+            lowerPrecedenceArg = type1;
         }
-
-        final boolean lowerPrecedenceCastable = lowerPrecedenceArg.isConvertableTo(higherPrecedenceArg, false);
-        final boolean higherPrecedenceCastable = higherPrecedenceArg.isConvertableTo(lowerPrecedenceArg, false);
-
-        if (lowerPrecedenceCastable) {
+        if (lowerPrecedenceArg.isConvertableTo(higherPrecedenceArg, false)) {
```

Comment:
Inlines these because if the first is true there is no need to evaluate the second.